File: macos_app/NagarikPatro/CalendarCore/Sources/CalendarCore/BsCalendarData.swift
import Foundation

/// Loads and provides access to the BS calendar month-length tables (282 years).
/// Thread-safe singleton. Shared between main app and WidgetKit extension.
public final class BsCalendarData: Sendable {
    public static let shared = BsCalendarData()

    public let meta: Meta
    public let monthNamesEn: [String]
    public let monthNamesNp: [String]
    public let weekdayNamesEn: [String]
    public let weekdayNamesNp: [String]
    public let weekdayNamesNpShort: [String]
    /// Key: BS year (Int). Value: [totalDays, month1..month12] — 13 elements.
    public let years: [Int: [Int]]

    public struct Meta: Sendable {
        public let source: String
        public let referenceAd: String  // "1913-04-13"
        public let referenceBsYear: Int
        public let referenceBsMonth: Int
        public let referenceBsDay: Int
        public let yearMin: Int
        public let yearMax: Int
        public let nepalTzOffsetSeconds: Int
    }

    private init() {
        guard let url = Bundle.module.url(forResource: "bs_calendar_data", withExtension: "json") else {
            fatalError("bs_calendar_data.json not found in CalendarCore bundle")
        }
        do {
            let data = try Data(contentsOf: url)
            let json = try JSONSerialization.jsonObject(with: data) as! [String: Any]

            let metaJson = json["meta"] as! [String: Any]
            let refBs = metaJson["reference_bs"] as! [String: Any]
            let yearRange = metaJson["year_range"] as! [Int]

            self.meta = Meta(
                source: metaJson["source"] as! String,
                referenceAd: metaJson["reference_ad"] as! String,
                referenceBsYear: refBs["year"] as! Int,
                referenceBsMonth: refBs["month"] as! Int,
                referenceBsDay: refBs["day"] as! Int,
                yearMin: yearRange[0],
                yearMax: yearRange[1],
                nepalTzOffsetSeconds: metaJson["nepal_tz_offset_seconds"] as! Int
            )

            self.monthNamesEn = json["month_names_en"] as! [String]
            self.monthNamesNp = json["month_names_np"] as! [String]
            self.weekdayNamesEn = json["weekday_names_en"] as! [String]
            self.weekdayNamesNp = json["weekday_names_np"] as! [String]
            self.weekdayNamesNpShort = json["weekday_names_np_short"] as! [String]

            let yearsJson = json["years"] as! [String: [Int]]
            var parsed: [Int: [Int]] = [:]
            parsed.reserveCapacity(yearsJson.count)
            for (key, value) in yearsJson {
                guard let year = Int(key) else { continue }
                parsed[year] = value
            }
            self.years = parsed
        } catch {
            fatalError("Failed to load bs_calendar_data.json: \(error)")
        }
    }

    /// Days in a specific month of a BS year. Returns nil if year/month is out of range.
    public func daysInMonth(year: Int, month: Int) -> Int? {
        guard month >= 1, month <= 12,
              let yearData = years[year],
              yearData.count == 13 else {
            return nil
        }
        return yearData[month]
    }

    /// Total days in a BS year. Returns nil if year is out of range.
    public func daysInYear(year: Int) -> Int? {
        years[year]?.first
    }

    /// Whether a BS year is in the supported range.
    public func isYearSupported(_ year: Int) -> Bool {
        years[year] != nil
    }
}

File: windows_app/NagarikPatro/Core/BsCalendarData.cs
using System.Text.Json;

namespace NagarikPatro.Core;

/// <summary>
/// Loads and provides access to the BS calendar month-length tables (282 years).
/// Thread-safe singleton.
/// </summary>
public sealed class BsCalendarData
{
    private static readonly Lazy<BsCalendarData> _instance = new(() => new BsCalendarData());
    public static BsCalendarData Instance => _instance.Value;

    public string Source { get; }
    public string ReferenceAd { get; }
    public int ReferenceBsYear { get; }
    public int ReferenceBsMonth { get; }
    public int ReferenceBsDay { get; }
    public int YearMin { get; }
    public int YearMax { get; }
    public int NepalTzOffsetSeconds { get; }

    public IReadOnlyList<string> MonthNamesEn { get; }
    public IReadOnlyList<string> MonthNamesNp { get; }
    public IReadOnlyList<string> WeekdayNamesEn { get; }
    public IReadOnlyList<string> WeekdayNamesNp { get; }
    public IReadOnlyList<string> WeekdayNamesNpShort { get; }

    /// <summary>
    /// Key: BS year. Value: int[13] where [0]=totalDays, [1..12]=days per month.
    /// </summary>
    public IReadOnlyDictionary<int, int[]> Years { get; }

    private BsCalendarData()
    {
        var path = Path.Combine(AppContext.BaseDirectory, "Data", "bs_calendar_data.json");
        if (!File.Exists(path))
            throw new FileNotFoundException($"bs_calendar_data.json not found at {path}");

        var json = File.ReadAllText(path);
        using var doc = JsonDocument.Parse(json);
        var root = doc.RootElement;

        var meta = root.GetProperty("meta");
        Source = meta.GetProperty("source").GetString()!;
        ReferenceAd = meta.GetProperty("reference_ad").GetString()!;
        var refBs = meta.GetProperty("reference_bs");
        ReferenceBsYear = refBs.GetProperty("year").GetInt32();
        ReferenceBsMonth = refBs.GetProperty("month").GetInt32();
        ReferenceBsDay = refBs.GetProperty("day").GetInt32();
        var yearRange = meta.GetProperty("year_range");
        YearMin = yearRange[0].GetInt32();
        YearMax = yearRange[1].GetInt32();
        NepalTzOffsetSeconds = meta.GetProperty("nepal_tz_offset_seconds").GetInt32();

        MonthNamesEn = ParseStringArray(root.GetProperty("month_names_en"));
        MonthNamesNp = ParseStringArray(root.GetProperty("month_names_np"));
        WeekdayNamesEn = ParseStringArray(root.GetProperty("weekday_names_en"));
        WeekdayNamesNp = ParseStringArray(root.GetProperty("weekday_names_np"));
        WeekdayNamesNpShort = ParseStringArray(root.GetProperty("weekday_names_np_short"));

        var yearsElement = root.GetProperty("years");
        var years = new Dictionary<int, int[]>();
        foreach (var prop in yearsElement.EnumerateObject())
        {
            if (!int.TryParse(prop.Name, out int year)) continue;
            var values = new int[prop.Value.GetArrayLength()];
            int i = 0;
            foreach (var v in prop.Value.EnumerateArray())
                values[i++] = v.GetInt32();
            years[year] = values;
        }
        Years = years;
    }

    public int? DaysInMonth(int year, int month)
    {
        if (month < 1 || month > 12) return null;
        if (!Years.TryGetValue(year, out var data) || data.Length != 13) return null;
        return data[month];
    }

    public int? DaysInYear(int year)
    {
        if (!Years.TryGetValue(year, out var data) || data.Length == 0) return null;
        return data[0];
    }

    public bool IsYearSupported(int year) => Years.ContainsKey(year);

    private static List<string> ParseStringArray(JsonElement element)
    {
        var list = new List<string>();
        foreach (var item in element.EnumerateArray())
            list.Add(item.GetString()!);
        return list;
    }
}
